explainers/bundle-urls-and-origins.md

```diff
@@ -0,0 +1,621 @@
+filtered *claimed URL* when the referrer's *bundle URL* matches the realm's
+*bundle URL*.
+
+> For example, if the true referrer is
+`package:https:,,distributor.example,package.wbn$https:,,claimed.example/page.html`,
+and the current document is
+`package:https:,,distributor.example,package.wbn$https:,,claimed.example/page2.html`,
+the `document.referrer` getter will notice the matching *bundle URLs* and return
+just the *claimed URL* out of the referrer. In this case, `document.referrer`
+is:
+>
+> ```url
+> https://claimed.example/page.html
+> ```
+>
+> Or if the current document is in a different bundle, `document.referrer` is:
+>
+> ```url
+> package:https:,,distributor.example,package.wbn$https:,,claimed.example/page.html
+> ```
+>
+> If the true referrer is `package:https:,,distributor.example,package.wbn` and
+the current document is in the same bundle, `document.referrer` is the empty
+string.
+
+To avoid adding exceptions to the [above algorithm](#referer-headers) for
+same-bundle navigations, there will be some cases where a same-bundle navigation
+has no referrer but the same navigation between resources on the web would have
+a referrer. Specifically, if the bundle is served with a referrer policy of
+`no-referrer`, `origin`, or `strict-origin`, the *claimed URL* is entirely
+removed from the referrer, so regardless of the subresource's referrer policy,
+`document.referrer` will be the empty string.
+
+### Origins
+
+#### `Origin` headers
+
+When a resource inside a bundle makes a cross-origin network request, what
+[`Origin`
+header](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Origin) should
+be sent?
+
+We propose the simple thing here: Compute the [referrer](#referer-headers) that
+would be used for a navigation to the requested resource and send the origin of
+that referrer string, computed by dropping the path.
+
+This omits [Fetch's
+allowance](https://fetch.spec.whatwg.org/#append-a-request-origin-header) for
+cors-mode and websocket requests to get the full origin even if the referrer
+policy would disallow it.
+
+As with the referrer, when the `Origin` header is observed within the same
+bundle, most likely via service workers, the bundle part needs to be removed.
+
+#### `postMessage` target origin
+
+When sending a message, the sender
+sets the
+[`targetOrigin`](https://developer.mozilla.org/en-US/docs/Web/API/Window/postMessage#Syntax)
+field to determine which recipients can read it.
+
+* When sending from outside the bundle to inside, or from inside one bundle to
+  inside a different bundle, the sender uses the full
+  `package:bundle-url$claimed-origin`.
+* When sending from inside the bundle to inside, the sender uses just the
+  claimed origin.
+* When sending from inside the bundle to outside, the sender uses the same
+  target origin that a sender outside the bundle would use. This means the
+  message could be intercepted by something inside the same bundle, which should
```

Comment:
Sorry, I don't understand correctly. Why the message could be intercepted by something inside the same bundle?

When "package:https:,,a.example.com$https:,,b.example/page.html" sends a message to "https://c.example/target.html", postMessage() should be called with "https://c.example" `targetOrigin`.
In this case, the message can't be intercepted anything inside the same bundle.
Is my understanding incorrect?

Reply:
My understanding writing this was that if `package:https:,,a.example.com$https:,,b.example/page.html` (a.k.a. "Packaged-B") has a `WindowProxy` whose name it sees as `https://c.example/target.html`, that `WindowProxy` could be referring to either the "real" `https://c.example/target.html` (Internet-C) fetched over the internet or `package:https:,,a.example.com$https:,,c.example/target.html` (Packaged-C) retrieved from inside the same bundle. And even if Packaged-B has some reason to prefer its message be received by Internet-C, it can't use `targetOrigin` to ensure that, because the two are expressed the same inside the `targetOrigin` field.

In the design as it's written here, Packaged-B has absolutely no way to tell those two situations apart (other than knowledge baked in when the package was created). I don't see any harm from giving it a way in the platform to tell the situations apart, but it adds some complication that I haven't been able to justify with a compensating benefit.

Reply:
I've added a similar example to the wording here.

Reply:
Thank you. Now I understand.